libraries/heft-config-file/src/ConfigurationFile.ts

```diff
@@ -300,36 +286,61 @@ export class ConfigurationFile<TConfigurationFile> {
     visitedConfigurationFilePaths: Set<string>,
     rigConfig: RigConfig | undefined
   ): Promise<TConfigurationFile> {
-    let cacheEntry: IConfigurationFileCacheEntry<TConfigurationFile> | undefined =
-      this._configurationFileCache.get(resolvedConfigurationFilePath);
-    if (!cacheEntry) {
-      try {
-        cacheEntry = {
-          configurationFile: await this._loadConfigurationFileInnerAsync(
-            terminal,
-            resolvedConfigurationFilePath,
-            visitedConfigurationFilePaths,
-            rigConfig
-          )
-        };
-      } catch (e) {
-        cacheEntry = { error: e };
-      }
-    } else {
-      terminal.writeVerboseLine(
-        `Found "${ConfigurationFile._formatPathForLogging(
-          resolvedConfigurationFilePath
-        )}" in ConfigurationFile path.`
+    let cacheEntryPromise: Promise<IConfigurationFileCacheEntry<TConfigurationFile>> | undefined =
+      this._configPromiseCache.get(resolvedConfigurationFilePath);
+    if (!cacheEntryPromise) {
+      cacheEntryPromise = this._createCacheEntryPromise(
+        this._loadConfigurationFileInnerAsync(
+          terminal,
+          resolvedConfigurationFilePath,
+          visitedConfigurationFilePaths,
+          rigConfig
+        )
       );
+      this._configPromiseCache.set(resolvedConfigurationFilePath, cacheEntryPromise);
     }
 
+    // We check for loops after caching a promise for this config file, but before attempting
+    // to resolve the promise. We can't handle loop detection in the `InnerAsync` function, because
+    // we could end up waiting for a cached promise (like A -> B -> A) that never resolves.
+    if (visitedConfigurationFilePaths.has(resolvedConfigurationFilePath)) {
```

Comment:
This can be `else if`, since it cannot be in the visited array without a cache entry having already existed.

Reply:
I agree with the logic, but readability-wise I'm not sure.

Right now the function is basically 3 snippets -- get or set a promise cache entry, perform loop detection, and then resolve or reject a value.  I see the "perform loop detection" snippet more as a safeguard for the last section (ensuring we don't accidentally wait on a promise that will never resolve), not so much related to the first snippet (get or set promise cache entry).

Turning it into an else-if merges the first and second snippet in a way that I think might be more confusing than helpful (but I guess it does save a function call in some situations)?

Reply:
It saves a lookup on the first read of every file. Granted, the lookup is a very minor cost relative to the file system reads, so it's a pretty minor performance gain.